File: 2_medium/0001_1000/00002_add_two_numbers/solution.cpp
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* res_dummy = new ListNode(0);
        ListNode* res_prev = res_dummy;

        int offset = 0;
        while (true) {

            int num1 = 0;
            if (l1 != nullptr) {
                num1 = l1->val;
                l1 = l1->next;
            }

            int num2 = 0;
            if (l2 != nullptr) {
                num2 = l2->val;
                l2 = l2->next;
            }

            int num_res = num1 + num2 + offset;
            offset = 0;
            if (num_res >= 10) {
                offset = 1;
                num_res -= 10;
            } 

            ListNode* new_node = new ListNode(num_res);
            res_prev->next = new_node;
            res_prev = new_node;

            if (l1 == nullptr && l2 == nullptr) {
                if (offset > 0) {
                    ListNode* new_node = new ListNode(offset);
                    res_prev->next = new_node;
                }

                break;
            } 
        }

        ListNode* res_head = res_dummy->next;
        delete res_dummy;
        return res_head;
    }
};

File: 2_medium/0001_1000/00002_add_two_numbers/solution.rs
impl Solution {
    pub fn add_two_numbers(mut l1: Option<Box<ListNode>>, mut l2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut res_dummy: Option<Box<ListNode>> = Some(Box::new(ListNode::new(0)));
        let mut res_prev = &mut res_dummy;
        let mut offset = 0;
        loop {
            let mut num1 = 0_i32;
            if l1.is_some() {
                let mut l1_node = l1.unwrap();
                num1 = l1_node.val;
                l1 = l1_node.next.take();
            }

            let mut num2 = 0_i32;
            if l2.is_some() {
                let mut l2_node = l2.unwrap();
                num2 = l2_node.val;
                l2 = l2_node.next.take();
            }

            let mut res_num = num1 + num2 + offset;
            offset = 0;
            if res_num >= 10 {
                offset = 1;
                res_num -= 10;
            }

            let mut new_node: Option<Box<ListNode>> = Some(Box::new(ListNode::new(res_num)));
            res_prev.as_mut().unwrap().next = new_node;
            res_prev = &mut res_prev.as_mut().unwrap().next;

            if l1.is_none() && l2.is_none() {
                if offset > 0 {
                    let mut new_node: Option<Box<ListNode>> = Some(Box::new(ListNode::new(offset)));
                    res_prev.as_mut().unwrap().next = new_node;
                }

                break;
            }
        }
        
        res_dummy.unwrap().next.take()
    }
}
